backend/debug/inlined_frame_ranges.ml

```diff
@@ -0,0 +1,189 @@
+ * DEALINGS IN THE SOFTWARE.                                                  *
+ ******************************************************************************)
+
+module L = Linear
+
+(* CR mshinwell/xclerc/poechsel:
+
+   1. maybe "set" isn't the correct algebraic structure for the Compute_ranges
+   interface
+
+   2. maybe a list-based structure could be used here instead (not for variable
+   ranges) *)
+
+module Inlined_frames = struct
+  module Key = struct
+    module T0 = struct
+      type t = Debuginfo.t
+
+      let print = Debuginfo.print_compact_extended
+
+      let compare t1 t2 =
+        (* It needs to be the case that [compare shorter longer] returns -1 (and
+           vice-versa) when [shorter] has fewer frames than [longer]. See
+           dwarf_inlined_frames:create_up_to_root. *)
+        let items1 = Debuginfo.to_items t1 in
+        let items2 = Debuginfo.to_items t2 in
+        let rec loop (items1 : Debuginfo.item list)
+            (items2 : Debuginfo.item list) =
+          match items1, items2 with
+          | [], [] -> 0
+          | [], _ :: _ -> -1
+          | _ :: _, [] -> 1
+          | i1 :: items1, i2 :: items2 ->
+            (* CR mshinwell: think about the [None] cases some more *)
+            let c = Option.compare String.compare i1.dinfo_uid i2.dinfo_uid in
+            if c <> 0
+            then c
+            else
+              let c =
+                Option.compare String.compare i1.dinfo_function_symbol
+                  i2.dinfo_function_symbol
+              in
+              if c <> 0 then c else loop items1 items2
+        in
+        loop items1 items2
+
+      let equal t1 t2 = compare t1 t2 = 0
+
+      let hash = Hashtbl.hash
+
+      let output _ _ = Misc.fatal_error "Not implemented"
+    end
+
+    include T0
+    include Identifiable.Make (T0)
+
+    type key = t
+
+    module Raw_set = Set
+
+    module Set = struct
+      type t =
+        | Ok of Raw_set.t
+        | Unreachable
+
+      let of_list keys = Ok (Raw_set.of_list keys)
+
+      let union t1 t2 =
+        match t1, t2 with
+        | Unreachable, _ | _, Unreachable -> Unreachable
+        | Ok s1, Ok s2 -> Ok (Raw_set.union s1 s2)
+
+      let inter t1 t2 =
+        match t1, t2 with
+        | Unreachable, t | t, Unreachable -> t
+        | Ok s1, Ok s2 -> Ok (Raw_set.inter s1 s2)
+
+      let diff t1 t2 =
+        match t1, t2 with
+        | Unreachable, _ -> Unreachable
+        | _, Unreachable -> Ok Raw_set.empty
+        | Ok s1, Ok s2 -> Ok (Raw_set.diff s1 s2)
+
+      let fold f t init =
+        match t with Unreachable -> init | Ok s -> Raw_set.fold f s init
+
+      let print ppf t =
+        match t with
+        | Unreachable -> Format.pp_print_string ppf "Unreachable"
+        | Ok s -> Raw_set.print ppf s
+    end
+
+    let parent t =
+      match List.rev t with [] | [_] -> None | _ :: t -> Some (List.rev t)
+
+    let all_parents _t = []
```

Comment:
This question is based mainly on
the function names, but shouldn't
that return a list with a least the
value returned by `parent t`?

Reply:
This returns the empty list because there might be a bug in `Compute_ranges`, see the CR there.  This was only an optimization anyway.